rfc/rfc-51/rfc-51.md

```diff
@@ -148,20 +152,27 @@ hudi_cdc_table/
 
 Under a partition directory, the `.log` file with `CDCBlock` above will keep the changing data we have to materialize.
 
-There is an option to control what data is written to `CDCBlock`, that is `hoodie.table.cdc.supplemental.logging`. See the description of this config above.
+#### Write-on-indexing vs Write-on-compaction
```

Comment:
This is an interesting trade-off. I understand Compaction is the one common place today where we know for sure the op type for all supported indexing. I am not sure about marrying 2 different concepts here - CDC stream freshness and Query efficiency (compaction frequency). This could be confusing for customers who dont know the implementation details. 

I was thinking along the lines of a new table service, ChangeTrackLoggingService, we can then
- Asynchronous cdc option in addition
      - no cdc related ingestion failures and no performance hit. 
- Optional for other indexes, but mandatory for bucket indexing when cdc is on
- Cant be frequently than compaction - lightweight when supplimental.logging is off. determines the CDC stream freshness.

This also means a different timeline instant to manage. I will think more on this and wait for suggestions as well. 

In general I am not a big fan of supplimental.logging=true - I think duplicating data values is something we should avoid completely. I dont even know how to deal with data inconsistencies between cdc blocks and data blocks if we end up in that state. 

cc @xushiyan @danny0405 @YannByron @vinothchandar

Reply:
> This could be confusing for customers who dont know the implementation details.
I agree with this. we should try to keep the behavior of hudi cdc consistent with customers's perception, and provide explanations only where inconsistent (for ex. bucket index)

Adding the new table service `ChangeTrackLoggingService`, that is worth thinking about. And customers also need to know this, know when they can deal with the hudi cdc stream.

About supplimental.logging=true, let us provide the two modes to deal with situations where customers are more concerned about query efficiency or write efficiency.

Reply:
> This could be confusing for customers who dont know the implementation details.

I actually don't resonate with this, as I don't see this as exposing implementation details. It's a known characteristic that RO queries for MOR tables depends on compaction. This aligns with MOR semantics: if you need to query efficiently (RO), you have to wait for compaction; if you need to get fresher results (RT/snapshot), you have to spend more computation power. Similarly, if you need to query CDC result efficiently, you wait for compaction; if to get fresher CDC results, spend more computation power to diff the log files. Logic for the 2nd case has not been proposed (and it should be) in the PR.

I feel having a new service for this would introduce more impl. complexities and gaps to understand. For example, the way it interplays with what type of index is used and whether cdc is on or not. We should note that in COW compaction happens implicitly, which would simplify the impl. for cdc write-on-compaction due to code path reuse.

As this only affects MOR tables, we should trace back to the MOR semantics: what can users expect to get from MOR?  my understanding is: MOR is the tradeoff btw fast ingestion and query freshness, it applies to the data files, it can also be applied to the cdc data.

Reply:
@xushiyan @prasannarajaperumal 
If we want to make this cdc behavior similar to the characteristic of RO/RT for MOR tables, provide a config like `hoodie.datasource.write.cdc.mor.strategy` that have two enum values: realtime and compaction. When `hoodie.datasource.write.cdc.mor.strategy` = `realtime`, we give the cdc result by merging the log files in-flight asap. if it's `compaction`, generate the cdc log files when compaction. in this `compaction` case, user can get the cdc result until compaction is finished.

Reply:
Some questions need to be cleared here:
1. when user wants to read the change log of  commits: c1, c2, c3, but the c1 and c2 are cleaned/archived, what is the behavior/semantics of hudi for both cow and mor table ?
2. when reading in streaming mode, and because of the consumer lagging, some commits requested are cleaned, what is the behavior, show we throws directly ?

Reply:
The valid range of cdc query is same to `timetravel`'s, only the instants in active timeline can be queried.

1. Only c3 can be queried. But if c3 needs to load the previous file slice that has been cleaned, throw exceptions. Because part of cdc data from c3 is lost.
2.  Unless the files that is needed to be loaded instants in current active timeline have already been cleaned, no exception will be thrown. That means some cdc data from the cleaned instants will be lost. The behavior is like that sync the binlog from mysql to kafka. If binlog is archived before they are synced to kafka, they can't be seen in kafka.

Reply:
Let us agree conceptually in the RFC the CDC stream consistency can be synchronous or asynchronous to the write. Freshness of this stream can be tied to compaction to begin with (potentially having a new background service if required in future). If we ensure that the CDC log writing logic is well abstracted and shared between the upsert and compaction we should be easily extend this to another service in future.

Reply:
OK, but one thing need to be noticed if persist mor's cdc data when compaction. @prasannarajaperumal @xushiyan 
give an example first: a record(id=1, name=x1) in base file, at t1 commit update  name to x2 (in logFile1), at t2 commit update to x3(in logFile2). CDC should return two changing records (x1->x2, x2->x3).
The current compaction implement will call `HoodieMergedLogRecordScanner` to get the log records first, then finish the compaction by `HoodieMergeHandler`. But the log records from `HoodieMergedLogRecordScanner` have already combined in advance, so that we lost some cdc info.
So if we wanna persist cdc when compaction for mor tables, we have to upgrade these related coded: `HoodieMergedLogRecordScanner` to make it return non-combined records, `HoodieCompactor` and `HoodieMergeHandler` to adapt these changes.

Reply:
Yes - HoodieMergedLogRecordScanner needs to store CDC state on each merged row which could be used in writing the CDC block after compaction. I think we should mention this explicitly in the design - It is okay to track this as a seperate jira and get back to it after the initial implementation. 

```Compaction will produce a row for every update recorded in the log files```

Reply:
+1 for this. i also have discussed with @xushiyan that a separate jira to track that write cdc when compaction. and provide a config to indicate that extract cdc data from log files in a flight way or cdc files which are written when compaction.

Reply:
@xushiyan @YannByron - Lets link the jira once its created.

Reply:
linked in "rollout" section